/*========================== begin_copyright_notice ============================

Copyright (C) 2016-2021 Intel Corporation

SPDX-License-Identifier: MIT

============================= end_copyright_notice ===========================*/

#include <cm/cm.h>

_GENX_MAIN_
void test() {
}

#define STRING2(x) #x
#define STRING(x) STRING2(x)

#ifdef CM_GENX
#pragma message ( "CM_GENX defined with value " STRING(CM_GENX) ) // expected-warning{{CM_GENX defined with value 970}}
#else
#warning CM_GENX not defined
#endif

#ifdef CM_GEN7_5
#warning CM_GEN7_5 defined
#endif

#ifdef CM_GEN8
#warning CM_GEN8 defined
#endif

#ifdef CM_GEN8_5
#warning CM_GEN8_5 defined
#endif

#ifdef CM_GEN9
#warning CM_GEN9 defined
#endif

#ifdef CM_GEN9_5
#warning CM_GEN9_5 defined // expected-warning{{CM_GEN9_5 defined}}
#endif

#ifdef CM_GEN10
#warning CM_GEN10 defined
#endif

#ifdef CM_GEN10_5
#warning CM_GEN10_5 defined
#endif

#ifdef CM_GEN11
#warning CM_GEN11 defined
#endif

// We test a number of different ways to specify a GLK jit target option.
// All are equivalent, and should produce the same results, so we only need
// one set of CHECK values for all of these tests.
// We check the expected Gen variant macros are defined (and no others), and
// that the Finalizer is called with the expected platform option.
// We also check that the expected files are generated by deleting them, 
// which also leaves things tidy for the next test.

// RUN: %cmc -emit-llvm -Qxcm_jit_target=glk -Xclang -verify -Xclang -verify-ignore-unexpected -- %s

// RUN: %cmc -emit-llvm -Qxcm_jit_target:glk -Xclang -verify -Xclang -verify-ignore-unexpected -- %s

// RUN: %cmc -emit-llvm -Qxcm_jit_targetglk -Xclang -verify -Xclang -verify-ignore-unexpected -- %s

// RUN: %cmc -emit-llvm -Qxcm_jit_target=gen9_5lp -Xclang -verify -Xclang -verify-ignore-unexpected -- %s

// RUN: %cmc -emit-llvm -Qxcm_jit_target=GEN9P5LP -Xclang -verify -Xclang -verify-ignore-unexpected -- %s

// RUN: %cmc -emit-llvm /Qxcm_jit_target=GLK -Xclang -verify -Xclang -verify-ignore-unexpected -- %s

// RUN: %cmc -emit-llvm /Qxcm_jit_target:glK -Xclang -verify -Xclang -verify-ignore-unexpected -- %s

// RUN: %cmc -emit-llvm /Qxcm_jit_targetGlk -Xclang -verify -Xclang -verify-ignore-unexpected -- %s

// RUN: %cmc -emit-llvm /Qxcm_jit_targetgLk -Xclang -verify -Xclang -verify-ignore-unexpected -- %s

// RUN: %cmc -emit-llvm -Qxcm_jit_target -mcpu=GLK -Xclang -verify -Xclang -verify-ignore-unexpected -- %s

// RUN: %cmc -emit-llvm -Qxcm_jit_target -mcpu=gEN9p5lP -Xclang -verify -Xclang -verify-ignore-unexpected -- %s

// RUN: %cmc -emit-llvm -march=Gen9p5LP -Qxcm_jit_target -Xclang -verify -Xclang -verify-ignore-unexpected -- %s

// RUN: %cmc -emit-llvm -Qxcm_jit_target -march=Gen9_5LP -Xclang -verify -Xclang -verify-ignore-unexpected -- %s

