/*========================== begin_copyright_notice ============================

Copyright (C) 2015-2021 Intel Corporation

SPDX-License-Identifier: MIT

============================= end_copyright_notice ===========================*/

#include <cm/cm.h>

_GENX_MAIN_
void test() {
}

#define STRING2(x) #x
#define STRING(x) STRING2(x)

#ifdef CM_GENX
#pragma message ( "CM_GENX defined with value " STRING(CM_GENX) ) // expected-warning{{CM_GENX defined with value 850}}
#else
#warning CM_GENX not defined
#endif

#ifdef CM_GEN7_5
#warning CM_GEN7_5 defined
#endif

#ifdef CM_GEN8
#warning CM_GEN8 defined
#endif

#ifdef CM_GEN8_5
#warning CM_GEN8_5 defined // expected-warning{{CM_GEN8_5 defined}}
#endif

#ifdef CM_GEN9
#warning CM_GEN9 defined
#endif

#ifdef CM_GEN9_5
#warning CM_GEN9_5 defined
#endif

#ifdef CM_GEN10
#warning CM_GEN10 defined
#endif

#ifdef CM_GEN10_5
#warning CM_GEN10_5 defined
#endif

#ifdef CM_GEN11
#warning CM_GEN11 defined
#endif

// We test a number of different ways to specify a CHV jit target option.
// All are equivalent, and should produce the same results, so we only need
// one set of CHECK values for all of these tests.
// We check the expected Gen variant macros are defined (and no others), and
// that the Finalizer is called with the expected platform option.
// We also check that the expected files are generated by deleting them, 
// which also leaves things tidy for the next test.

// RUN: %cmc -emit-llvm -Qxcm_jit_target=chv -Xclang -verify -Xclang -verify-ignore-unexpected -- %s

// RUN: %cmc -emit-llvm -Qxcm_jit_target:chv -Xclang -verify -Xclang -verify-ignore-unexpected -- %s

// RUN: %cmc -emit-llvm -Qxcm_jit_targetchv -Xclang -verify -Xclang -verify-ignore-unexpected -- %s

// RUN: %cmc -emit-llvm -Qxcm_jit_target=gen8lp -Xclang -verify -Xclang -verify-ignore-unexpected -- %s

// RUN: %cmc -emit-llvm -Qxcm_jit_target=GEN8LP -Xclang -verify -Xclang -verify-ignore-unexpected -- %s

// RUN: %cmc -emit-llvm -Qxcm_jit_target=Gen8LP -Xclang -verify -Xclang -verify-ignore-unexpected -- %s

// RUN: %cmc -emit-llvm -Qxcm_jit_target=GEN8_5 -Xclang -verify -Xclang -verify-ignore-unexpected -- %s

// RUN: %cmc -emit-llvm -Qxcm_jit_target=gen8_5 -Xclang -verify -Xclang -verify-ignore-unexpected -- %s

// RUN: %cmc -emit-llvm -Qxcm_jit_target=gen8_5 -Xclang -verify -Xclang -verify-ignore-unexpected -- %s

// RUN: %cmc -emit-llvm /Qxcm_jit_target=Gen8_5 -Xclang -verify -Xclang -verify-ignore-unexpected -- %s

// RUN: %cmc -emit-llvm /Qxcm_jit_target:chV -Xclang -verify -Xclang -verify-ignore-unexpected -- %s

// RUN: %cmc -emit-llvm /Qxcm_jit_targetChv -Xclang -verify -Xclang -verify-ignore-unexpected -- %s

// RUN: %cmc -emit-llvm /Qxcm_jit_targetcHv -Xclang -verify -Xclang -verify-ignore-unexpected -- %s

// RUN: %cmc -emit-llvm -Qxcm_jit_target -mcpu=CHV -Xclang -verify -Xclang -verify-ignore-unexpected -- %s

// RUN: %cmc -emit-llvm -Qxcm_jit_target -mcpu=Gen8Lp -Xclang -verify -Xclang -verify-ignore-unexpected -- %s

// RUN: %cmc -emit-llvm -Qxcm_jit_target -mcpu=Gen8_5 -Xclang -verify -Xclang -verify-ignore-unexpected -- %s

// RUN: %cmc -emit-llvm -march=Chv -Qxcm_jit_target -Xclang -verify -Xclang -verify-ignore-unexpected -- %s

// RUN: %cmc -emit-llvm -Qxcm_jit_target -march=Gen8lP -Xclang -verify -Xclang -verify-ignore-unexpected -- %s

